### PR TITLE
Enhancement

### PR DESCRIPTION
**Pull request by chengchingwen**
The PR adds more complete support of the safetensors format:
1. support serialization
2. support `BFloat16` and `Float8`
3. allow lazy loading with `mmap`
4. handle endian correctly
5. "safe"r (according to `huggingface/safetensors`)

Most of the implementation follows [the rust code](https://github.com/huggingface/safetensors/tree/main/safetensors), so it should be easier to track any changes or updates.

The main unexported APIs are `serialize`, `deserialize`, and `SafeTensor`, see the updated readme.

cc @pevnak, @ToucheSir 

### PR Checklist

- [x] Tests are added
- [x] Documentation, if applicable

### CODE REVIEW

**Comment by chengchingwen** [comment]
seems like there isn't any coverage tracker enabled for this repo?

File: README.md [hunk 1]
```diff
@@ -45,9 +41,69 @@ Dict{String, Array} with 27 entries:
...
+julia> tensors = SafeTensors.deserialize("test/model.safetensors"; mmap = true #= default to `true`=#);
+
+julia> tensors["float32_35"]
+3×5 mappedarray(ltoh, PermutedDimsArray(reshape(reinterpret(Float32, view(::Vector{UInt8}, 0x0000000000000ef5:0x0000000000000f30)), 5, 3), (2, 1))) with eltype Float32:
+  0.0   1.0   2.0   3.0   4.0
+  5.0   6.0   7.0   8.0   9.0
+ 10.0  11.0  12.0  13.0  14.0
+```
+
+Serialization is also support:
```
Comment (CarloLucibello) [inline README.md:55]:
```suggestion
Serialization is also supported:
```

File: README.md [hunk 2]
```diff
@@ -45,9 +41,69 @@ Dict{String, Array} with 27 entries:
...
+
 ```julia
-julia> d = SafeTensors.load_header("model.safetensors")
+julia> weights = Dict("W"=>randn(BFloat16, 3, 5), "b"=>rand(BFloat16, 3))
```
Comment (CarloLucibello) [inline README.md:58]:
```suggestion
julia> using Random, BFloat16s

julia> weights = Dict("W"=>randn(BFloat16, 3, 5), "b"=>rand(BFloat16, 3))
```

File: README.md [hunk 3]
```diff
@@ -45,9 +41,69 @@ Dict{String, Array} with 27 entries:
...
+Dict{String, Array{BFloat16}} with 2 entries:
+  "W" => [0.617188 0.695312 … 0.390625 -2.0; -0.65625 -0.617188 … 0.652344 0.244141; 0.226562 2.70312 … -0.174805 -0.7773…
+  "b" => [0.111816, 0.566406, 0.283203]
+
+julia> f = tempname();
+
+julia> SafeTensors.serialize(f, weights)
+
+julia> loaded["W"] ≈ weights["W"]
+true
```
Comment (CarloLucibello) [inline README.md:68]: this should be moved down  I guess

Comment (chengchingwen) [reply README.md:70]: It's just for showing that the result is correct

Comment (chengchingwen) [reply README.md:72]: I miss a line of deserialize, fixed

**Review by CarloLucibello — APPROVED** [review]

**Comment by chengchingwen** [comment]
bump

**Pull request merged**